### PR TITLE
SensorPlayer Quick Fix

### PR DESCRIPTION
**Pull request by VivaanBahl**
This is an issue where the encoder subscribers can't yet handle fingerprint messages. This was the only thing wrong I found during testing

### CODE REVIEW

**Comment by TrevorDecker** [comment]
so your saying that this is ready for merge?

**Comment by VivaanBahl** [comment]
yes. Otherwise playback will error out. This has been tested.

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/RobobuggyMainFile.java [hunk 1]
```diff
@@ -43,24 +43,21 @@ public static void main(String[] args)  {
...
     		//Play back mode enabled
-    		final SensorPlayer sp = new SensorPlayer("logs/sensors_2016-02-05-20-03-47.txt");
-        	new Thread(new Runnable() {
-				
-				@Override
-				public void run() {
-					sp.run();
-				}
-			}).start();
+    		SensorPlayer sp = new SensorPlayer("logs/2016-02-12-21-52-18/sensors_2016-02-12-21-52-18.txt");
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/RobobuggyMainFile.java]: Codacy Issue found: [Avoid unused local variables such as 'sp'.](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2039563468/issues/source?bid=2938665&fileBranchId=3128705#l46)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/simulation/SensorPlayer.java [hunk 2]
```diff
@@ -57,34 +56,50 @@
...
-	@Override
-	public void run() {
+
+    public void setupPlaybackTrigger() {
+        new Subscriber(NodeChannel.GUI_LOGGING_BUTTON.getMsgPath(), new MessageListener() {
+            @Override
+            public void actionPerformed(String topicName, Message m) {
+                GuiLoggingButtonMessage message = (GuiLoggingButtonMessage)m;
+                if (message.getLoggingMessage().equals(GuiLoggingButtonMessage.LoggingMessage.START)) {
+                    if (!SensorPlayer.this.isAlive()) {
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/simulation/SensorPlayer.java]: Codacy Issue found: [These nested if statements could be combined](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2039563512/issues/source?bid=2938665&fileBranchId=3128705#l93)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/simulation/SensorPlayer.java [hunk 3]
```diff
@@ -96,108 +111,124 @@ public void run() {
...
+                        long currentSensorTime = timeMessage.getTimestamp().getTime();
+                        long timeDiff = currentSensorTime - prevSensorTime;
+                        if (timeDiff > 10) {
+                            Thread.sleep(timeDiff);
                         }
+                        prevSensorTime = currentSensorTime;
+                    }
+
+
+                    switch (versionID) {
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/simulation/SensorPlayer.java]: Codacy Issue found: [Switch statements should have a default label](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2039563512/issues/source?bid=2938665&fileBranchId=3128705#l208)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/simulation/SensorPlayer.java [hunk 4]
```diff
@@ -29,25 +29,24 @@
 import com.roboclub.robobuggy.messages.StateMessage;
 import com.roboclub.robobuggy.messages.SteeringMeasurement;
 import com.roboclub.robobuggy.messages.WheelAngleCommandMeasurement;
-import com.roboclub.robobuggy.ros.Message;
-import com.roboclub.robobuggy.ros.Publisher;
-import com.roboclub.robobuggy.ros.NodeChannel;
+import com.roboclub.robobuggy.ros.*;
+import sun.management.Sensor;
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/simulation/SensorPlayer.java]: Codacy Issue found: [Avoid unused imports such as 'sun.management.Sensor'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2039563512/issues/source?bid=2938665&fileBranchId=3128705#l33)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/simulation/SensorPlayer.java [hunk 5]
```diff
@@ -96,108 +111,124 @@ public void run() {
...
 
-                        if(sensorTimeFromPrev > 0){
-                            Thread.sleep(sensorTimeFromPrev);
+                    if (prevSensorTime == -1) {
+                        prevSensorTime = timeMessage.getTimestamp().getTime();
+                    }
+                    else {
+                        long currentSensorTime = timeMessage.getTimestamp().getTime();
+                        long timeDiff = currentSensorTime - prevSensorTime;
+                        if (timeDiff > 10) {
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/simulation/SensorPlayer.java]: Codacy Issue found: [Avoid using Literals in Conditional Statements](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2039563512/issues/source?bid=2938665&fileBranchId=3128705#l201)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/simulation/SensorPlayer.java [hunk 6]
```diff
@@ -57,34 +56,50 @@
...
+                    if (!SensorPlayer.this.isAlive()) {
+                        SensorPlayer.this.start();
+                    }
+                }
+            }
+        });
+    }
+
+    @Override
+    public void run() {
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/simulation/SensorPlayer.java]: Codacy Issue found: [The method run() has an NPath complexity of 202](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2039563512/issues/source?bid=2938665&fileBranchId=3128705#l102)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/simulation/SensorPlayer.java [hunk 7]
```diff
@@ -57,34 +61,55 @@
...
+    /**
+     * Sets up the playback trigger - hitting the start button will go through 1 iteration of a log file
+     */
+    public void setupPlaybackTrigger() {
+        new Subscriber(NodeChannel.GUI_LOGGING_BUTTON.getMsgPath(), new MessageListener() {
+            @Override
+            public void actionPerformed(String topicName, Message m) {
+                GuiLoggingButtonMessage message = (GuiLoggingButtonMessage)m;
+                if (message.getLoggingMessage().equals(GuiLoggingButtonMessage.LoggingMessage.START)) {
+                    if (!SensorPlayer.this.isAlive()) {
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/simulation/SensorPlayer.java]: Codacy Issue found: [These nested if statements could be combined](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2041124211/issues/source?bid=2938665&fileBranchId=3128705#l103)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/simulation/SensorPlayer.java [hunk 8]
```diff
@@ -57,34 +61,55 @@
...
+                        SensorPlayer.this.start();
+                    }
+                }
+            }
+        });
+    }
+
+    @Override
+    public void run() {
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/simulation/SensorPlayer.java]: Codacy Issue found: [The method run() has an NPath complexity of 202](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2041124211/issues/source?bid=2938665&fileBranchId=3128705#l112)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/simulation/SensorPlayer.java [hunk 9]
```diff
@@ -96,108 +121,128 @@ public void run() {
...
 
-                        if(sensorTimeFromPrev > 0){
-                            Thread.sleep(sensorTimeFromPrev);
+                    if (prevSensorTime == -1) {
+                        prevSensorTime = timeMessage.getTimestamp().getTime();
+                    }
+                    else {
+                        long currentSensorTime = timeMessage.getTimestamp().getTime();
+                        long timeDiff = currentSensorTime - prevSensorTime;
+                        if (timeDiff > 10) {
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/simulation/SensorPlayer.java]: Codacy Issue found: [Avoid using Literals in Conditional Statements](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2041124211/issues/source?bid=2938665&fileBranchId=3128705#l213)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/LoggingPanel.java [hunk 10]
```diff
@@ -79,6 +85,22 @@ public void setFileName(String fileName) {
 	}
 
 	/**
+	 * safe returns the speed
+	 * @return the speed, or 1 if the text in the field can't be serialized
+	 */
+	public double getPlaybackSpeed() {
+		String speedStr = playbackSpeed.getText();
+		if (speedStr.equals("")) return 1;
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/LoggingPanel.java]: Codacy Issue found: [Avoid using if statements without curly braces](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2041124223/issues/source?bid=2938665&fileBranchId=3128705#l93)

Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/LoggingPanel.java]: Codacy Issue found: [Position literals first in String comparisons](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2041124223/issues/source?bid=2938665&fileBranchId=3128705#l93)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/Gui.java [hunk 11]
```diff
@@ -55,6 +55,7 @@ public static Gui getInstance() {
 
 	//The windowList is a list of all of the windows that are a part of the gui
 	private ArrayList<RobobuggyJFrame> windowList = new ArrayList<RobobuggyJFrame>();
+	private MainGuiWindow mainGuiWindow;
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/Gui.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2041124219/issues/source?bid=2938665&fileBranchId=3128705#l58)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/RobobuggyMainFile.java [hunk 12]
```diff
@@ -43,24 +43,26 @@ public static void main(String[] args)  {
...
+                new RobobuggyLogicNotification("Notification 1", RobobuggyMessageLevel.NOTE);
+                new RobobuggyLogicNotification("Notification 2", RobobuggyMessageLevel.NOTE);
+
+                Thread.sleep(500);
+
+                new RobobuggyLogicNotification("Notif 3", RobobuggyMessageLevel.NOTE);
+            }
+            catch (InterruptedException e) {
+                new RobobuggyLogicNotification("interreupted", RobobuggyMessageLevel.NOTE);
+            }
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/RobobuggyMainFile.java]: why do we need to sleep here? 

why are we sending 3 notifications?

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/simulation/SensorPlayer.java (hunk 7 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/simulation/SensorPlayer.java]: Can we do this

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/simulation/SensorPlayer.java (hunk 9 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/simulation/SensorPlayer.java]: Yeah lets move this to a final static setting set in config or at least the top of this file

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/GpsPanel.java [hunk 13]
```diff
@@ -107,7 +107,8 @@ public void paintComponent(Graphics g) {
 
 		g.drawImage(map, 0, 0, frameWidth, frameHeight, Color.black, null);
 
-		for	(LocTuple mTuple : locs) {
+		for (int i = 0; i < locs.size(); i++) {
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/GpsPanel.java]: why?   I would prefer to keep this as a for each, it allows the compiler to make things faster. Not that it really matters in this case

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/LoggingPanel.java (hunk 10 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/LoggingPanel.java]: so I was screwed once by not having braces around an if, can we please use them

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/simulation/SensorPlayer.java [hunk 14]
```diff
@@ -225,4 +270,11 @@ private boolean validateLogFileMetadata(JsonObject logFile) {
 
         return true;
     }
+
+    /**
+     * gets the new playback speed from the GUI and puts it into playbackSpeed
+     */
+    public void getNewPlaybackSpeed() {
+        playbackSpeed = Gui.getInstance().getMainGuiWindow().getCtrlPanel().getLoggingPanel().getPlaybackSpeed();
+    }
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/simulation/SensorPlayer.java]: this is super ugly can you change it to just making the loggingPanel a factory so you can directly reference it?

**Comment by TrevorDecker** [comment]
so I just had some style changes, everything else looks good.

Once you make the noted changes I am good for a merge

**Pull request merged**